aiida/engine/processes/process_spec.py

```diff
@@ -76,22 +78,36 @@ def exit_code(self, status, label, message, invalidates_cache=False):
 
         self._exit_codes[label] = ExitCode(status, message, invalidates_cache=invalidates_cache)
 
+    # override return type to aiida's PortNamespace subclass
+
+    @property
+    def ports(self) -> PortNamespace:
```

Comment:
Hmm, not sure about this. You have to change it here _because_ the port namespace type is changeable through the `PORT_NAMESPACE_TYPE` class attribute. If this were to be changed dynamically, the typing would again be wrong. How is this typically dealt with in `mypy`? You cannot do variable dereferencing I suppose, i.e. `-> 'ProcessSpec.PORT_NAMESPACE_TYPE'`?

Reply:
> You cannot do variable dereferencing

no, not to my knowledge

the point is though that `PORT_NAMESPACE_TYPE` can no longer be any subclass of the plumpy `PortNamespace`, because the aiida `Process` requires it to have a `serialize` method. So even if you dynamically change it, it still has to be a subclass of the aiida `PortNamespace`.

Reply:
yes, but even if you define another `PORT_NAMESPACE_TYPE` that is a subclass of `PortNamespace`, you also have to override the properties again just to get the type hinting correct? That seems like problematic behavior for a use case that should be quite common. Is this a known limitation of mypy at least? I won't make this stop this PR from being merged but would like to know if these kinds of designs will make type hinting painful

Reply:
>  Is this a known limitation of mypy at least?

Err well personally I would say it is more highlighting a "problematic" code design, than a limitation of mypy.
Its fine if you are just using `PORT_NAMESPACE_TYPE` to define different concrete implementations of the same abstract class (with the same methods), then you would not need these method overrides.
But when you are using it to define subclasses that do not adhere to the base class (e.g. adding new methods) I am not sure that is an ideal design for dependency injection.

Reply:
Wait, maybe I am misunderstanding, but we _are_ returning a subclass of the original `PortNamespace`, or is that not the problem? Surely that should be fine as a design. I agree that it should respect the same interface as a minimum, but why would it be bad to _add_ methods? I agree that removing them is a bad idea

Reply:
Because `aiida.engine.Process` has fundamentally changed the required interface of the `PortNamespace` (by requiring the `serialize` method), i.e. the`aiida.engine.Process` will fail if you set its `PORT_NAMESPACE_TYPE = plumpy.PortNamespace`.